src/pipelines/mod.rs

```diff
@@ -405,6 +406,75 @@ impl fmt::Display for AttrWriter<'_> {
     }
 }
 
+/// A wrapper for Html modifications, and rewrites.
+pub struct Document(String);
```

Comment:
I would prefer two changes here:

* use a `Vec<u8>`
* provide a constructor, maybe `new(data: impl Into<Vec<u8>>)`, or something can could accept `&String`, `String`, or `&str`. Whatever makes sense.

I don't think we need to turn this into a string every time. And I think it would be good to have some "finish now" method, which retrieves the final string.

I am not too worried about performance. But if this turns out to be a viable pattern, we might even be able to queue/batch those requests and only call them when calling that "finish" method.

Reply:
Storing a `Vec<u8>` makes more sense.  Will add a constructor with generics. :+1: 

I also have a few optimizations in mind (didn't test any of them), but I decided it would be better to focus on removing `nipper` since that's the PR's main goal. I think a queue system would make it harder to debug the code, so while I'm not a fan of that idea in particular, I think we could try to find other optimizations before going down that route.

Reply:
Awesome!

> I also have a few optimizations in mind (didn't test any of them), but I decided it would be better to focus on removing nipper since that's the PR's main goal.

I totally agree!

> I think a queue system would make it harder to debug the code, so while I'm not a fan of that idea in particular, I think we could try to find other optimizations before going down that route.

Good point. It might also make passing `fn`s much harder.

Reply:
> I don't think we need to turn this into a string every time. And I think it would be good to have some "finish now" method, which retrieves the final string.

Is there any reason we need a `String` at all? 
Any of the processing done on the Html, is done by passing `&[u8]`'s. 

In case we want to print it out, we can implement a method on `Document` such as:
```rust
fn as_str(&self) -> &str {
    std::str::from_utf8(self.0.as_slice()).expect("Some error handling.")
}
```
We can also just implement `Display` but since we already have some bytes we can get a `&str` without extra allocations.

Reply:
On another note, I have ditched the generic constructor. Since we store everything in a `Vec<u8>` inside `Document`, and not a `String` we can use `std::fs::read` instead of `std::fs::read_to_string`.

This approach is more direct, avoids an extra allocation, and supports more encodings. `std::fs::read_to_string` can return an encoding error if the content is not UTF-8 (UTF-8 encoding is recommended but not required from as part of the standard).